src/backend/commands/createas.c

```diff
@@ -609,28 +419,239 @@ ExecCreateTableAs(CreateTableAsStmt *stmt, const char *queryString,
 
 		if (into->ivm)
 		{
-
 			Oid matviewOid = address.objectId;
 			Relation matviewRel = table_open(matviewOid, NoLock);
 			Relids	relids = NULL;
 
-			copied_query = copyObject(query);
-			AcquireRewriteLocks(copied_query, true, false);
+			/*
+			 * Mark relisivm field, if it's a matview and into->ivm is true.
+			 */
+			SetMatViewIVMState(matviewRel, true);
 
-			CreateIvmTriggersOnBaseTables(copied_query, (Node *)copied_query->jointree, matviewOid, &relids);
+			if (!into->skipData)
+			{
+				copied_query = copyObject(query);
+				AcquireRewriteLocks(copied_query, true, false);
 
+				CreateIvmTriggersOnBaseTables(copied_query, (Node *)copied_query->jointree, matviewOid, &relids);
+				bms_free(relids);
+			}
 			table_close(matviewRel, NoLock);
-
-			bms_free(relids);
 		}
 	}
 
 	return address;
 }
+
+/*
+ * rewriteQueryForIMMV -- rewrite query for definition query of IMMV
+ * noerror is used by check_ivm_restriction_walker().
+ */
+Query *
+rewriteQueryForIMMV(Query *query, List *colNames, bool noerror)
+{
+	Query *rewritten;
+
+	TargetEntry *tle;
+	Node *node;
+	ParseState *pstate = make_parsestate(NULL);
+	FuncCall *fn;
+
+	check_ivm_restriction_context ctx = {false, false, false, NIL};
+	check_ivm_restriction_walker((Node *) query, &ctx, 0, noerror);
```

Comment:
I see. The check is not necessary. But... a reason of called this function, is that I get 'ctx' flags.
This flag is used for judging to contain the outer join. 

I think the best way is that separate off process of judging outer join.

Reply:
I got it. You mean that calling check_ivm_restriction_walker is necessary because ctx-has_outerjoin is checked afterwards, right?

So, what' needed here is to check if the view has outer joins instead of checking restrictions of IVM. We don't need to modify check_ivm_restriction_walker, I think.  The required check is simple because we don't allow nested queries with outer join for now, like a place in IVM_immediate_maintenance where hasOuterjoins flag is made, although I don't care you make a new function to check this.

Reply:
> I got it. You mean that calling check_ivm_restriction_walker is necessary because ctx-has_outerjoin is checked afterwards, right?
That's right.

Reply:
Thank you for your comment.
I am going to add simple check process when execute REFRESH, and remove noerror flag.